handler/oauth2/flow_token_exchange.go

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright © 2020 Bosch.IO, Germany
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @author		Miguel Paulos Nunes <Miguel.PaulosNunes@bosch.io>, Olaf Märker <Olaf.Maerker@bosch.io>
+ * @copyright 	2020 Bosch.IO, Germany
+ * @license 	Apache-2.0
+ *
+ */
+
+package oauth2
+
+import (
+	"context"
+	"github.com/ory/fosite/storage"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/ory/fosite"
+)
+
+// TokenExchangeGrantHandler is a response handler for the Token Exchange grant
+// as defined in https://tools.ietf.org/html/rfc8693
+type TokenExchangeGrantHandler struct {
+	AccessTokenStrategy      AccessTokenStrategy
+	AccessTokenStorage       AccessTokenStorage
+	AccessTokenLifespan      time.Duration
+	ScopeStrategy            fosite.ScopeStrategy
+	AudienceMatchingStrategy fosite.AudienceMatchingStrategy
+	RefreshTokenStrategy     RefreshTokenStrategy
+	RefreshTokenLifespan     time.Duration
+	RefreshTokenScopes       []string
+	CoreStrategy
+	CoreStorage
+	Store fosite.Storage
+}
+
+// HandleTokenEndpointRequest implements https://tools.ietf.org/html/rfc8693#section-2.1 (currently impersonation only)
+func (c *TokenExchangeGrantHandler) HandleTokenEndpointRequest(ctx context.Context, request fosite.AccessRequester) error {
```

Comment:
To me it looks like `actor_token` and `actor_token_type` are currently ignored. If they are not implemented, we should throw an error instead of silently accepting their values (and potentially returning a success). To be honest, I also don't really understand what the value of `actor_token` would be and neither its implications. But as far as I can tell those are values used when performing delegation, and this code currently only supports impersonation.

Reply:
> If they are not implemented, we should throw an error instead of silently accepting their values

I agree and will add the checks for that.

> To be honest, I also don't really understand what the value of actor_token would be and neither its implications

We had exactly the same problem. After several discussions on that we came up with a scenario for which it may make sense.
E.g. User A want to allow User B to access service D over service C. Assuming Service C has tokens of both users, it exchanges the token from B (actor) with that from A (subject) and calls D with the resulting token.
Unfortunately I could not find further information around to clarify that.